compiler/rustc_codegen_llvm/src/abi.rs

```diff
@@ -344,16 +344,20 @@ impl ArgAbiMethods<'tcx> for Builder<'a, 'll, 'tcx> {
 }
 
 pub trait FnAbiLlvmExt<'tcx> {
-    fn llvm_type(&self, cx: &CodegenCx<'ll, 'tcx>) -> &'ll Type;
+    fn llvm_type(&self, cx: &CodegenCx<'ll, 'tcx>, decl: bool) -> &'ll Type;
     fn ptr_to_llvm_type(&self, cx: &CodegenCx<'ll, 'tcx>) -> &'ll Type;
     fn llvm_cconv(&self) -> llvm::CallConv;
     fn apply_attrs_llfn(&self, cx: &CodegenCx<'ll, 'tcx>, llfn: &'ll Value);
     fn apply_attrs_callsite(&self, bx: &mut Builder<'a, 'll, 'tcx>, callsite: &'ll Value);
 }
 
 impl<'tcx> FnAbiLlvmExt<'tcx> for FnAbi<'tcx, Ty<'tcx>> {
-    fn llvm_type(&self, cx: &CodegenCx<'ll, 'tcx>) -> &'ll Type {
-        let args_capacity: usize = self.args.iter().map(|arg|
+    fn llvm_type(&self, cx: &CodegenCx<'ll, 'tcx>, decl: bool) -> &'ll Type {
+        // Ignore extra args when calling C variadic functions.
```

Comment:
Finally another thing I find confusing here is that we tell LLVM the function as variadic regardless of `decl`, so I'm not exactly sure what purpose this serves? Is it intended to remove the `VaList` style argument?

Reply:
Say you have `fn(i32, i64, ...)`, and you call it like `fn(i32, i64, bool, f32)`, I think the last two argument types need to be removed to make sure that the function signature at the call site matches the declaration. Maybe always using `&self.args[..self.fixed_count]` would work too though?

Reply:
So, this variadic thing was the last bit I struggled with, and I'm happy to explore better ways to represent it.

The specific example that tripped when building `std` was `libc::fcntl`, which looks like this:

```rust
pub fn fcntl(fd: ::c_int, cmd: ::c_int, ...) -> ::c_int;
```

That turns into LLVM `fn(i32, i32, ...) -> i32`. The call site for `set_cloexec` in `std` passes three args, and the `FnAbi::llvm_type` from there gave me `fn(i32, i32, i32, ...) -> i32` -- having that "extra" arg when its `FnAbi` was constructed, but still represented as variadic. However, that tripped an assertion:

```
rustc: [...]/rust/src/llvm-project/llvm/lib/Bitcode/Writer/ValueEnumerator.h:167: unsigned int llvm::ValueEnumerator::getTypeID(llvm::Type*) const: Assertion `I != TypeMap.end() && "Type not in ValueEnumerator!"' failed.
```

So I used `fixed_count` to try and get the original declaration type, which seems to work -- giving the same signature as `element_type(val_ty(fn_ptr))` would if we didn't care about being opaque.

> Maybe always using `&self.args[..self.fixed_count]` would work too though?

It at least needs to check for `c_variadic`, because I know that the `#[track_caller]` argument for `&Location` is also past the `fixed_count`. I can try it without considering the new `decl` parameter though, or outright removing that.

Reply:
> It at least needs to check for c_variadic, because I know that the #[track_caller] argument for &Location is also past the fixed_count.

That seems bad. At least on riscv that will cause the `#[track_caller]` argument to be handled like a vararg during the abi calculation.

Reply:
> At least on riscv that will cause the `#[track_caller]` argument to be handled like a vararg during the abi calculation.

I looked at that, but it's only called by `adjust_for_cabi`, which isn't used for Rust/intrinsic ABIs.

Reply:
All tests passed locally without using `decl`, even with LLVM assertions, so I've removed it.